pp_sort.c: new build-time warning when compiling with gcc-8

Compiling with gcc-8.3.0 on Linux, two previously unobserved warnings have emerged.

```
$ git describe
v5.31.9-106-g85e384a62b

$ ./perl -Ilib -V:config_args
config_args='-des -Dusedevel -Dcc=gcc-8 -Duseithreads';

$ gcc-8 --version | head -1
gcc-8 (Ubuntu 8.3.0-26ubuntu1~18.04) 8.3.0

$ parse-build-warnings 85e384a62b.linux.gcc-8.threaded.make.output.gz 
File:  85e384a62b.linux.gcc-8.threaded.make.output.gz

[
  {
    char   => 74,
    group  => "Wunused-parameter",
    line   => 338,
    source => "pp_sort.c",
    text   => "unused parameter \xE2\x80\x98flags\xE2\x80\x99",
  },
  {
    char   => 17,
    group  => "Wcast-function-type",
    line   => 240,
    source => "Opcode.xs",
    text   => "cast between incompatible function types from \xE2\x80\x98void (*)(PerlInterpreter *, const char *, ...)\xE2\x80\x99 {aka \xE2\x80\x98void (*)(struct interpreter *, const char *, ...)\xE2\x80\x99} to \xE2\x80\x98void (*)(void *)\xE2\x80\x99",
  },
]
```

In context:

```
gcc-8 -c -DPERL_CORE -D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/us
r/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -std=c89 -O2 -Wall -Werror=pointer-arith -Wextra -Wc++
-compat -Wwrite-strings -Werror=declaration-after-statement pp_sort.c
pp_sort.c: In function ‘S_sortsv_flags_impl’:
pp_sort.c:338:74: warning: unused parameter ‘flags’ [-Wunused-parameter]
 S_sortsv_flags_impl(pTHX_ gptr *base, size_t nmemb, SVCOMPARE_t cmp, U32 flags)
                                                                      ~~~~^~~~~
```
```
gcc-8 -c   -D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/in
clude -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -Wall -Werror=pointer-arith -Wextra -Wc++-compat -Wwrite-strings
 -Werror=declaration-after-statement -O2   -DVERSION=\"1.45\" -DXS_VERSION=\"1.45\" -fPIC "-I../.."   Opcode.c
In file included from ../../perl.h:5579,
                 from Opcode.xs:3:
Opcode.xs: In function ‘opmask_addlocal’:
Opcode.xs:240:17: warning: cast between incompatible function types from ‘void (*)(PerlInterpreter *, const char *, ...)’ {aka ‘void (*)(struct interpreter *, const char *, ...)’} to ‘void (*)(void *)’ [-Wcast-function-type]
  SAVEDESTRUCTOR((void(*)(void*))Perl_warn,"PL_op_mask restored");
                 ^
../../embed.h:496:57: note: in definition of macro ‘save_destructor’
 #define save_destructor(a,b) Perl_save_destructor(aTHX_ a,b)
                                                         ^
Opcode.xs:240:2: note: in expansion of macro ‘SAVEDESTRUCTOR’
  SAVEDESTRUCTOR((void(*)(void*))Perl_warn,"PL_op_mask restored");
  ^~~~~~~~~~~~~~
rm -f ../../lib/auto/Opcode/Opcode.so
gcc-8  -shared -O2 -L/usr/local/lib -fstack-protector-strong  Opcode.o  -o ../../lib/auto/Opcode/Opcode.so  \
      \
  
chmod 755 ../../lib/auto/Opcode/Opcode.so
```

The source location reported for the `pp_sort.c` warning received a commit on March 09:
```
commit 044d25c73ce10d2b29008b875209d414303ccff7
Author:     Tomasz Konojacki <me@xenu.pl>
AuthorDate: Tue Mar 3 00:45:05 2020 +0100
Commit:     Karl Williamson <khw@cpan.org>
CommitDate: Mon Mar 9 07:55:49 2020 -0600

    optimize sort by inlining comparison functions
    
    This makes special-cased forms such as sort { $b <=> $a }
    even faster.
    
    Also, since this commit removes PL_sort_RealCmp, it fixes the
    issue with nested sort calls mentioned in gh #16129
```
@xenu, @khwilliamson : Can you take a look?

The source location reported for the Opcode.xs warning was last touched in 1997 -- so something as yet unidentified must have changed recently to trigger this warning.

Thank you very much.
Jim Keenan

> Compiling with gcc-8.3.0 on Linux, two previously unobserved warnings have emerged.
> 
> ```
> $ git describe
> v5.31.9-106-g85e384a62b
> 
> $ ./perl -Ilib -V:config_args
> config_args='-des -Dusedevel -Dcc=gcc-8 -Duseithreads';
> 
> $ gcc-8 --version | head -1
> gcc-8 (Ubuntu 8.3.0-26ubuntu1~18.04) 8.3.0
> 
> $ parse-build-warnings 85e384a62b.linux.gcc-8.threaded.make.output.gz 
> File:  85e384a62b.linux.gcc-8.threaded.make.output.gz
> 
> [
>   {
>     char   => 74,
>     group  => "Wunused-parameter",
>     line   => 338,
>     source => "pp_sort.c",
>     text   => "unused parameter \xE2\x80\x98flags\xE2\x80\x99",
>   },
>   {
>     char   => 17,
>     group  => "Wcast-function-type",
>     line   => 240,
>     source => "Opcode.xs",
>     text   => "cast between incompatible function types from \xE2\x80\x98void (*)(PerlInterpreter *, const char *, ...)\xE2\x80\x99 {aka \xE2\x80\x98void (*)(struct interpreter *, const char *, ...)\xE2\x80\x99} to \xE2\x80\x98void (*)(void *)\xE2\x80\x99",
>   },
> ]
> ```
> 
> In context:
> 
> ```
> gcc-8 -c -DPERL_CORE -D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/us
> r/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -std=c89 -O2 -Wall -Werror=pointer-arith -Wextra -Wc++
> -compat -Wwrite-strings -Werror=declaration-after-statement pp_sort.c
> pp_sort.c: In function ‘S_sortsv_flags_impl’:
> pp_sort.c:338:74: warning: unused parameter ‘flags’ [-Wunused-parameter]
>  S_sortsv_flags_impl(pTHX_ gptr *base, size_t nmemb, SVCOMPARE_t cmp, U32 flags)
>                                                                       ~~~~^~~~~
> ```
> 
> ```
> gcc-8 -c   -D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/in
> clude -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -Wall -Werror=pointer-arith -Wextra -Wc++-compat -Wwrite-strings
>  -Werror=declaration-after-statement -O2   -DVERSION=\"1.45\" -DXS_VERSION=\"1.45\" -fPIC "-I../.."   Opcode.c
> In file included from ../../perl.h:5579,
>                  from Opcode.xs:3:
> Opcode.xs: In function ‘opmask_addlocal’:
> Opcode.xs:240:17: warning: cast between incompatible function types from ‘void (*)(PerlInterpreter *, const char *, ...)’ {aka ‘void (*)(struct interpreter *, const char *, ...)’} to ‘void (*)(void *)’ [-Wcast-function-type]
>   SAVEDESTRUCTOR((void(*)(void*))Perl_warn,"PL_op_mask restored");
>                  ^
> ../../embed.h:496:57: note: in definition of macro ‘save_destructor’
>  #define save_destructor(a,b) Perl_save_destructor(aTHX_ a,b)
>                                                          ^
> Opcode.xs:240:2: note: in expansion of macro ‘SAVEDESTRUCTOR’
>   SAVEDESTRUCTOR((void(*)(void*))Perl_warn,"PL_op_mask restored");
>   ^~~~~~~~~~~~~~
> rm -f ../../lib/auto/Opcode/Opcode.so
> gcc-8  -shared -O2 -L/usr/local/lib -fstack-protector-strong  Opcode.o  -o ../../lib/auto/Opcode/Opcode.so  \
>       \
>   
> chmod 755 ../../lib/auto/Opcode/Opcode.so
> ```
> 
> The source location reported for the `pp_sort.c` warning received a commit on March 09:
> 
> ```
> commit 044d25c73ce10d2b29008b875209d414303ccff7
> Author:     Tomasz Konojacki <me@xenu.pl>
> AuthorDate: Tue Mar 3 00:45:05 2020 +0100
> Commit:     Karl Williamson <khw@cpan.org>
> CommitDate: Mon Mar 9 07:55:49 2020 -0600
> 
>     optimize sort by inlining comparison functions
>     
>     This makes special-cased forms such as sort { $b <=> $a }
>     even faster.
>     
>     Also, since this commit removes PL_sort_RealCmp, it fixes the
>     issue with nested sort calls mentioned in gh #16129
> ```
> 
> @xenu, @khwilliamson : Can you take a look?
> 
> The source location reported for the Opcode.xs warning was last touched in 1997 -- so something as yet unidentified must have changed recently to trigger this warning.
> 
> Thank you very much.
> Jim Keenan

Further investigation suggests that the `cast-function-type` warning being generated from `ext/Opcode/Opcode.xs` has "always" been there -- where, by "always" is meant "since we began using gcc-8 to build perl".  I.e., gcc-8 generates a warning here where (some) older compilers did not.

So this is not a recent regression.  I'll open a separate ticket to explore this warning.  From this point forward, this ticket will only be concerned with the `unused parameter` warning in `pp_sort.c`.

Thank you very much.
Jim Keenan

--- a/pp_sort.c
+++ b/pp_sort.c
@@ -347,6 +347,7 @@ S_sortsv_flags_impl(pTHX_ gptr *base, size_t nmemb, SVCOMPARE_t cmp, U32 flags)
     gptr *which[3];
     off_runs stack[60], *stackp;
 
+    PERL_UNUSED_ARG(flags);
     PERL_ARGS_ASSERT_SORTSV_FLAGS_IMPL;
     if (nmemb <= 1) return;                     /* sorted trivially */
 

